File: Exercises/RMA/F-MPI-RMA/traffic-sm.f90
program traffic

  use iso_c_binding, only: c_ptr, c_f_pointer

  use mpi

  use trafficlib

  implicit none

! Set the size of the road

  integer :: ncell = 100000
  integer :: maxiter, printfreq

  integer :: rank, size, ierr, nlocal, rankup, rankdown, tag = 1
  integer, dimension(MPI_STATUS_SIZE) :: status

  integer :: i, iter, nmove, nmovelocal, ncars
  real    :: density

  integer, allocatable, dimension(:) :: newroad, bigroad
  integer, pointer, dimension(:) :: oldroadtmp, oldroad

  double precision :: tstart, tstop

  ! RMA variables

  integer :: nodecomm
  integer :: nodesize, noderank, nodestringlen
  integer :: nodewin
  integer(MPI_ADDRESS_KIND) :: winsize
  integer :: intsize, disp_unit
  character*(MPI_MAX_PROCESSOR_NAME) :: nodename
  type(c_ptr) :: baseptr

  
  maxiter = 200000000/ncell
  printfreq = maxiter/10

! Set target density of cars

  density = 0.52

  write(*,*) 'Running message-passing traffic model'

! Start message passing system

  call MPI_Init(ierr)

! Compute size and rank

  call MPI_Comm_size(MPI_COMM_WORLD, size, ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  nlocal = ncell/size

  if (rank == 0) then

     write(*,*)
     write(*,*) 'Length of road is ', ncell
     write(*,*) 'Number of iterations is ', maxiter
     write(*,*) 'Target density of cars is ', density
     write(*,*) 'Running on ', size, ' processes'

  end if

! Allocate arrays

  allocate(newroad(0:nlocal+1))
!  allocate(oldroad(0:nlocal+1))

  ! Create node-local communicator

  call MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, rank, &
                           MPI_INFO_NULL, nodecomm, ierr)

  ! Check it all went as expected
  
  call MPI_Get_processor_name(nodename, nodestringlen, ierr)
  call MPI_Comm_size(nodecomm, nodesize, ierr)
  call MPI_Comm_rank(nodecomm, noderank, ierr)

  write(*,*) 'Rank ', rank,' in COMM_WORLD is rank ', noderank, &
             ' in nodecomm on node ', nodename(1:nodestringlen)

  call MPI_Type_size(MPI_INTEGER, intsize, ierr)

  ! Allocated oldroad as a shared array, contiguous across processes

  winsize = (nlocal+2)*intsize

  ! displacements counted in units of integers

  disp_unit = intsize

  call MPI_Win_allocate_shared(winsize, disp_unit, &
       MPI_INFO_NULL, nodecomm, baseptr, nodewin, ierr)

  ! coerce baseptr to a Fortran array

  call c_f_pointer(baseptr, oldroadtmp, [nlocal+2])

  ! make sure indexing starts at 0 not at 1

  oldroad(0:nlocal+1) => oldroadtmp(1:nlocal+2)
  
  do i = 1, nlocal
     oldroad(i) = 0
     newroad(i) = 0
  end do

  if (rank == 0) then

     allocate(bigroad(ncell))

     ! Initialise road accordingly using random number generator

     write(*,*) 'Initialising ...'

     ncars = initroad(bigroad, ncell, density, seed)

     write(*,*) '... done'

     write(*,*) 'Actual density of cars is ', float(ncars)/float(ncell)
     write(*,*) 'Scattering data ...'

  end if

  call MPI_Scatter(bigroad,    nlocal, MPI_INTEGER, &
                   oldroad(1), nlocal, MPI_INTEGER, &
                   0, MPI_COMM_WORLD, ierr)

  if (rank == 0) then

     write(*,*) '... done'
     write(*,*)

  end if

! Compute neighbours

  rankup   = mod(rank + 1, size)
  rankdown = mod(rank + size - 1, size)

  tstart = MPI_Wtime()

  do iter = 1, maxiter

     ! Implement halo swaps which now includes boundary conditions

     ! Use direct memory write for on-node comms

     ! Ensure everyone is ready
     
     call MPI_Win_fence(0, nodewin, ierr)

     ! Write last cell (entry nlocal) to lower halo (entry 0)
     ! of up neighbour which is entry nlocal+2 of local array

     if (noderank /= nodesize-1) then

        oldroad(nlocal+2) = oldroad(nlocal)

     end if

     ! Write first cell (entry 1) to upper halo (entry nlocal+1)
     ! of down neighbour which is entry -1 of local array

     if (noderank /= 0) then

        oldroad(-1) = oldroad(1)

     end if

     ! Ensure everyone is finished before accessing oldroad

     call MPI_Win_fence(0, nodewin, ierr)

     ! Now do standard MPI communications extreme end points

     if (noderank == nodesize-1) then

        ! Send upper boundary to and receive upper halo from rankup

        call MPI_Ssend(oldroad(nlocal),  1, MPI_INTEGER, rankup, tag, &
		       MPI_COMM_WORLD, ierr)

        call MPI_Recv(oldroad(nlocal+1), 1, MPI_INTEGER, rankup, tag, &
                      MPI_COMM_WORLD, status, ierr)
     end if

     if (noderank == 0) then

        ! Send lower boundary to and receive lower halo from rankdown

        call MPI_Recv(oldroad(0), 1, MPI_INTEGER, rankdown, tag, &
                      MPI_COMM_WORLD, status, ierr)

	call MPI_Ssend(oldroad(1), 1, MPI_INTEGER, rankdown, tag, &
                       MPI_COMM_WORLD, ierr)

      end if



     nmovelocal = updateroad(newroad, oldroad, nlocal)

! Globally sum the value

     call MPI_Allreduce(nmovelocal, nmove, 1, MPI_INTEGER, MPI_SUM, &
                        MPI_COMM_WORLD, ierr)

! Copy new to old array

     do i = 1, nlocal

        oldroad(i) = newroad(i)

     end do

     if (mod(iter, printfreq) == 0) then

        if (rank == 0) write(*,*) 'At iteration ', iter, &
             ' average velocity is ', float(nmove)/float(ncars)
     end if

  end do

  tstop = MPI_Wtime()
  
  deallocate(newroad)

  if (rank == 0) then

     deallocate(bigroad)

     write(*,*)
     write(*,*) 'Finished'
     write(*,*)
     write(*,*) 'Time taken was  ', tstop - tstart, ' seconds'
     write(*,*) 'Update rate was ', &
                 1.d-6*float(ncell)*float(maxiter)/(tstop-tstart), ' MCOPs'
     write(*,*)

  end if

  call MPI_Finalize(ierr)

end program traffic

File: Exercises/RMA/C-MPI-RMA/traffic-sm.c
#include <stdlib.h>
#include <stdio.h>

#include <mpi.h>

#include "traffic.h"

int main(int argc, char **argv)
{
  // Set the size of the road

  int ncell = 100000;

  int *oldroad, *newroad, *bigroad;

  int i, iter, nmove, nmovelocal, ncars;
  int maxiter, printfreq; 

  float density; 

  double tstart, tstop;

  MPI_Status status;
  int rank, size, nlocal, rankup, rankdown;
  int tag = 1;

  // RMA variables

  MPI_Comm nodecomm;
  int nodesize, noderank, nodestringlen, upoffset, dnoffset;
  MPI_Win nodewin;
  MPI_Aint winsize;
  int disp_unit;
  char nodename[MPI_MAX_PROCESSOR_NAME];

  maxiter = 200000000/ncell; 
  printfreq = maxiter/10; 

  // Set target density of cars

  density = 0.52;

  printf("Running message-passing traffic model\n");

  // Start MPI

  MPI_Init(NULL, NULL);  

  // Find size and rank

  MPI_Comm_size(MPI_COMM_WORLD, &size);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  if (size == 1)
    {
      printf("Error: naive method for MPI halo-swapping requires more than 1 process\n");
      MPI_Finalize();
      return 1;
    }

  if (rank == 0)
    {
      printf("\nLength of road is %d\n", ncell);
      printf("Number of iterations is %d \n", maxiter);
      printf("Target density of cars is %f \n", density);
      printf("Running on %d processes\n", size);
    }

  nlocal = ncell/size;

  // Create node-local communicator

  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, rank,
              MPI_INFO_NULL, &nodecomm);

  // Check it all went as expected
  
  MPI_Get_processor_name(nodename, &nodestringlen);
  MPI_Comm_size(nodecomm, &nodesize);
  MPI_Comm_rank(nodecomm, &noderank);

  printf("Rank %d in COMM_WORLD is rank %d in nodecomm on node <%s>\n",
	 rank, noderank, nodename);

  //  oldroad = (int *) malloc((nlocal+2)*sizeof(int));
  newroad = (int *) malloc((nlocal+2)*sizeof(int));

  // Allocated oldroad as a shared array, contiguous across processes

  winsize = (nlocal+2)*sizeof(int);

  // displacements counted in units of integers

  disp_unit = sizeof(int);

  MPI_Win_allocate_shared(winsize, disp_unit,
              MPI_INFO_NULL, nodecomm, &oldroad, &nodewin);

  for (i=1; i <= nlocal; i++)
    {
      oldroad[i] = 0;
      newroad[i] = 0;
    }

  if (rank == 0)
    {
      bigroad = (int *) malloc(ncell*sizeof(int));

      // Initialise road accordingly using random number generator

      printf("Initialising road ...\n");
  
      ncars = initroad(bigroad, ncell, density, SEED);

      printf("...done\n");
      printf("Actual density is %f\n", (float) ncars / (float) ncell);
      printf("Scattering data ...\n");
    }

  MPI_Scatter(bigroad,     nlocal, MPI_INT,
	      &oldroad[1], nlocal, MPI_INT,
	      0, MPI_COMM_WORLD);

  if (rank == 0)
    {
      printf("... done\n\n");
    }

  // Compute neighbours

  rankup   = (rank + 1) % size;
  rankdown = (rank + size - 1) % size;

  tstart = MPI_Wtime();

  for (iter=1; iter<=maxiter; iter++)
    { 
      // Implement halo swaps which now includes boundary conditions

      // Use direct memory write for on-node comms

      // Ensure everyone is ready

      MPI_Win_fence(0, nodewin);

      // Write last cell (entry nlocal) to lower halo (entry 0)
      // of up neighbour which is entry nlocal+2 of local array

      if (noderank != nodesize-1)
	{
	  oldroad[nlocal+2] = oldroad[nlocal];
	}

      // Write first cell (entry 1) to upper halo (entry nlocal+1)
      // of down neighbour which is entry -1 of local array

      if (noderank != 0)
	{
	  oldroad[-1] = oldroad[1];
	}

      // Ensure everyone is finished before accessing oldroad

      MPI_Win_fence(0, nodewin);

      // Now do standard MPI communications extreme end points

      if (noderank == nodesize-1)
	{
	  // Send upper boundary to and receive upper halo from rankup

	  MPI_Ssend(&oldroad[nlocal],  1, MPI_INT, rankup, tag,
		   MPI_COMM_WORLD);

	  MPI_Recv(&oldroad[nlocal+1], 1, MPI_INT, rankup, tag,
		    MPI_COMM_WORLD, &status);
	}

      if (noderank == 0)
	{
	  // Send lower boundary to and receive lower halo from rankdown

	  MPI_Recv(&oldroad[0],     1, MPI_INT, rankdown, tag,
		    MPI_COMM_WORLD, &status);

	  MPI_Ssend(&oldroad[1],    1, MPI_INT, rankdown, tag,
		   MPI_COMM_WORLD);
	}

      // Apply CA rules to all cells

      nmovelocal = updateroad(newroad, oldroad, nlocal);

      // Globally sum the value

      MPI_Allreduce(&nmovelocal, &nmove, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);

      // Copy new to old array

      for (i=1; i<=nlocal; i++)
	{
	  oldroad[i] = newroad[i]; 
	}

      if (iter%printfreq == 0)
	{
	  if (rank == 0)
	    {
	      printf("At iteration %d average velocity is %f \n",
		     iter, (float) nmove / (float) ncars);
	    }
	} 
    }

  tstop = MPI_Wtime();

  //  free(oldroad);

  MPI_Win_free(&nodewin);
  free(newroad);

  if (rank == 0)
    {
      free(bigroad);

      printf("\nFinished\n");
      printf("\nTime taken was  %f seconds\n", tstop-tstart);
      printf("Update rate was %f MCOPs\n\n", \
      1.e-6*((double) ncell)*((double) maxiter)/(tstop-tstart));
    }

  // Finish

  MPI_Finalize();
}
